### PR TITLE
Introduce non-persistent topic and tests

### PR DESCRIPTION
**Pull request by rdhabalia**
### Motivation

As discussed at #451 : introducing non-persistent topic.

### Modifications

**Changes**:
Broker will support non-persistent topic to support above usecase. Broker doesn’t persist messages for non-persistent topic but it immediately delivers message to already connected consumers and replicators. It also supports partitioned-topic and global-topic. Broker replicates non-persistent messages to configured replicated cluster..

**Client Changes:**
Client can create producer and consumer on non-persistent topic using similar API as persistent-topic. However, non-persistent topic name protocol starts with `non-persistent` instead `persistent`. 
e.g.
`non-persistent://my-property/cluster/namespace/topicName`
**changes**:
- Client provides topic name that starts with `non-persistent` prefix.
- Broker keep pushing messages to consumer as soon as producer publishes so, if consumer is not fast enough to consume messages and internal receive queue is full with the received messages then client consumer will drop newly incoming messages.. Therefore, client must tune consumer receive-queue size based on consume rate.

**Broker Changes:**
- Broker creates a special non-persistent topic entity which manages non-persistent topic’s producers, consumers and replication.
- Broker will not cache or persist published message and try to immediately dispatch messages to connected subscribers and replicators.
- **Rate limiting:** Broker must be configurable with rate-limiting to limit rate of publish messages on each topic to prevent over usage of n/w bandwidth by a specific topic. Broker fails published message once publish rate goes beyond rate-limiting.

### Result

No existing functional changes.

### CODE REVIEW

File: pulsar-common/src/main/java/com/yahoo/pulsar/common/naming/DestinationDomain.java [hunk 1]
```diff
@@ -16,5 +16,29 @@
 package com.yahoo.pulsar.common.naming;
 
 public enum DestinationDomain {
-    persistent
+    persistent("persistent"), non_persistent("non-persistent");
```
Comment (merlimat) [inline pulsar-common/src/main/java/com/yahoo/pulsar/common/naming/DestinationDomain.java]: @rdhabalia Can you check whether there are any problems in using `non-persistent` in a URL schema? Is the `-` a valid character there? 
If there is a problem there, we could use `nonpersistent` or `memory`

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/non/persistent/NonPersistentDispatcher.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
+/**
```
Comment (saandrews) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/non/persistent/NonPersistentDispatcher.java]: Can we name the path as service/nonpersistent/ as one word in stead of two levels non/persistent ?

File: pulsar-common/src/main/java/com/yahoo/pulsar/common/naming/DestinationDomain.java (hunk 1 above)
Comment (rdhabalia) [inline pulsar-common/src/main/java/com/yahoo/pulsar/common/naming/DestinationDomain.java]: yes, `non-persistent` is a valid token for URL schema. We already have it at [NonPersistentTopics](https://github.com/rdhabalia/pulsar/blob/34534370d7bb71fe83175a818090f951824696f2/pulsar-broker/src/main/java/com/yahoo/pulsar/broker/admin/NonPersistentTopics.java#L51) and tested with testcase. However, I think we can change it to `nonpersistent` for consistency as `non-persistent` is not a valid variable/enum name. So, if we don't have any concern then I will change it to `nonpersistent`

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/non/persistent/NonPersistentDispatcher.java (hunk 2 above)
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/non/persistent/NonPersistentDispatcher.java]: yes, I will fix it.

File: pulsar-common/src/main/java/com/yahoo/pulsar/common/naming/DestinationDomain.java (hunk 1 above)
Comment (merlimat) [inline pulsar-common/src/main/java/com/yahoo/pulsar/common/naming/DestinationDomain.java]: > We already have it at NonPersistentTopics and tested with testcase.

But int that case it's part of the path, right? Like `http://localhost:8080/admin/non-persistent/.....`

The topic itself would be `non-persistent://prop/cluster/namespace/topic`.

I don't have preference for any of the 2 names, though. Just curious whether that is a valid protocol scheme.

Comment (rdhabalia) [inline pulsar-common/src/main/java/com/yahoo/pulsar/common/naming/DestinationDomain.java]: > But int that case it's part of the path, right? Like http://localhost:8080/admin/non-persistent/.....

Thant's correct, as [HttpLookupService](https://github.com/yahoo/pulsar/blob/master/pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/HttpLookupService.java#L72) creates path with given `destination-name(non-persistent://...)` and it is getting served by [NonPersistentTopics. getPartitionedMetadata servlet](https://github.com/rdhabalia/pulsar/blob/34534370d7bb71fe83175a818090f951824696f2/pulsar-broker/src/main/java/com/yahoo/pulsar/broker/admin/NonPersistentTopics.java#L63)

**Comment by rdhabalia** [comment]
as per discussion at #451 made two changes:
1. Broker dispatches message to subscriber as per available permits else broker will drop message for that subscriber and doesn't deliver the message if that subscriber doesn't have permit.
2. Broker allows configured number of max in-flight messages per connection and beyond that limit, broker will not process that message and immediately sends `success-ack` back to producer.
**Note:** by default broker sets `msgId=0:0` except on above usecase2 where if message is not processed successfully then it sends `msgId=-1:-1` (at least some indication that broker has not processed the message successfully due to rate limit)

**Comment by rdhabalia** [comment]
Closing this PR as all the changes are merged in #538

**Pull request closed without merging**